spec/features/groups/request_upgrade_spec.rb

```diff
@@ -0,0 +1,52 @@
+require "rails_helper"
+
+feature "Request an upgrade for a group", type: :feature do
+  let!(:group) do
+    create(:group, organisation: editor_user.organisation).tap do |group|
+      create(:membership, user: editor_user, group:, role: :group_admin)
+    end
+  end
+
+  scenario "a group admin requests an upgrade that is approved by an admin user" do
+    login_as_editor_user
+
+    visit groups_path
+    expect(page.find("h1")).to have_text "Your groups"
+    expect_page_to_have_no_axe_errors(page)
+
+    click_link group.name
+    expect(page.find("h1")).to have_text group.name
+    expect(page).to have_css ".govuk-caption-l", text: "Trial group"
+    expect_page_to_have_no_axe_errors(page)
+
+    click_link "Find out how to upgrade this group so you can make forms live"
+    expect(page.find("h1")).to have_text "Request to upgrade this trial group"
+    expect_page_to_have_no_axe_errors(page)
+
+    click_link_or_button "Send request to upgrade"
+    expect(page.find("h1")).to have_text "Your upgrade request has been sent"
+    expect_page_to_have_no_axe_errors(page)
+
+    login_as_organisation_admin_user
+
+    visit groups_path
+    expect(page).to have_css ".govuk-notification-banner", text: "You have one request to upgrade a trial group."
+    expect_page_to_have_no_axe_errors(page)
```

Comment:
On most other feature specs we follow the pattern of having private methods that have names describing each step of the test.

For this test in particular, we would end up with names like `i_visit_the_groups_page_as_an_organisation_admin` as we have to repeat the same actions as different users. We thought this made it harder follow the test with no benefit of reducing duplication so didn't follow this pattern.

Thoughts welcome on this.

Reply:
I think thats sensible 👍

Reply:
I think another option would be to have generic methods (e.g. `i_visit_the_groups_page`) which do the page visiting and the axe check, but keep the bits that change between roles outside of the method.

```ruby
def i_visit_the_groups_page
  visit groups_path
  expect_page_to_have_no_axe_errors(page)
end
```

and in the scenario:
```ruby
    login_as_editor_user
    i_visit_the_groups_page
    expect(page.find("h1")).to have_text "Your groups"
    
    login_as_organisation_admin_user
    i_visit_the_groups_page
    expect(page).to have_css ".govuk-notification-banner", text: "You have one request to upgrade a trial group."
```

Additionally I think you could also try putting the role-specific assertions into their own methods - wouldn't do anything to reduce duplication, but would potentially make the scenarios easier to read:

```ruby
def i_should_see_the_your_groups_text
  expect(page.find("h1")).to have_text "Your groups"
end

def i_should_see_the_upgrade_request_banner
  expect(page.find("h1")).to have_text "Your groups"
end
```

so the scenario would look like this:
```ruby
    login_as_editor_user
    i_visit_the_groups_page
    i_should_see_the_your_groups_text    

    login_as_organisation_admin_user
    i_visit_the_groups_page
    i_should_see_the_upgrade_request_banner
```

Reply:
I'll leave it up to you whether you think that's worthwhile - I don't think this spec is particularly difficult to read anyway

Reply:
Thanks for the suggestions. What you've suggested makes sense to me. In absence of a strong opinion, I think we'll leave it as it is because as you say it's not particularly difficult to read.

I've put in a topic in the dev meeting next week about this just to see if anyone has any strong preferences or good reasoning for using this style of having methods in the tests.